### PR TITLE
Add Terraform code to support provisioning of DNS and Images AWS accounts

### PR DESCRIPTION
**Pull request by jsf9k**
## 🗣 Description

This pull request adds Terraform code to support the provisioning of the Audit, DNS, Images, Log Archive, and Master AWS accounts in the COOL via assumption of roles.

## 💭 Motivation and Context

I want to deploy OpenVPN and FreeIPA to the Shared Services account in the COOL, which means that I need to first deploy some [cisagov/certboto-docker](https://github.com/cisagov/certboto-docker) infrastructure to the COOL DNS account.  Since I needed to bootstrap that account, I figured I may as well bootstrap the other accounts as well.

## 🧪 Testing

I deployed these changes to production as a means of testing.  I also verified that the pre-commit hooks all pass.

## 🚥 Types of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (causes existing functionality to change)

## ✅ Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by jsf9k** [comment]
@dav3r, please have a look at the changes in e66034c and 8a674ce.  (We don't need to configure the pettifogger0 account, but I added it to provide an example of using the Terraform module to reduce the amount of duplicate code - particularly in the `users` subdirectory.)

Note that setting up the pettifogger0 account required _no_ additional code to be added to the `users` subdirectory, which is a great improvement.

If you like this technique then I will change the other subdirectories to use it tomorrow.

**Comment by dav3r** [comment]
> @dav3r, please have a look at the changes in [e66034c](https://github.com/cisagov/cool-accounts/commit/e66034c919b26e86a1f7751e9a645878c6c82c02) and [8a674ce](https://github.com/cisagov/cool-accounts/commit/8a674ce463af1c43c93a0d7cf6c60645996fb80b). (We don't need to configure the pettifogger0 account, but I added it to provide an example of using the Terraform module to reduce the amount of duplicate code - particularly in the `users` subdirectory.)
> 
> Note that setting up the pettifogger0 account required _no_ additional code to be added to the `users` subdirectory, which is a great improvement.
> 
> If you like this technique then I will change the other subdirectories to use it tomorrow.

@jsf9k That module is excellent- I'm 100% on board with doing it this way. 🎯

**Comment by jsf9k** [comment]
> In some places we have used `ProvisionAccount` and in other places we have used `${var.provisionaccount_role_name}` (e.g. [here](https://github.com/cisagov/cool-accounts/blob/develop/terraform/provider.tf#L5) and [here](https://github.com/cisagov/cool-accounts/blob/develop/users/provider.tf#L5)). We need to decide if we are hard-coding `ProvisionAccount` across the board or if we are giving folks the option to configure that role name in each account.

Fixed in ca77013.  An oversight on my part.

**Comment by jsf9k** [comment]
> Typo: "furthered" should just be "further".

Fixed in f98a816.

File: audit/README.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+   ```
+
+1. Create a Terraform workspace (if you haven't already done so) by running
+   `terraform workspace new <workspace_name>`
+1. Create a `<workspace_name>.tfvars` file with all of the required
+   variables (see [Inputs](#Inputs) below for details):
+
+   ```console
+   this_account_id = "111111111111"
+   user_account_id = "222222222222"
```
Comment (dav3r) [inline audit/README.md]: This should be `users_account_id` to match the variable name

File: dns/README.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+   ```
+
+1. Create a Terraform workspace (if you haven't already done so) by running
+   `terraform workspace new <workspace_name>`
+1. Create a `<workspace_name>.tfvars` file with all of the required
+   variables (see [Inputs](#Inputs) below for details):
+
+   ```console
+   this_account_id = "111111111111"
+   user_account_id = "222222222222"
```
Comment (dav3r) [inline dns/README.md]: `users_account_id` (again; I'll stop mentioning these- you know what to do)

File: pettifogger0/README.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+   ```
+
+1. Create a Terraform workspace (if you haven't already done so) by running
+   `terraform workspace new <workspace_name>`
+1. Create a `<workspace_name>.tfvars` file with all of the required
+   variables (see [Inputs](#Inputs) below for details):
+
+   ```console
+   this_account_id = "111111111111"
+   user_account_id = "222222222222"
```
[the remark this answers is not in the document]
Comment (dav3r) [reply pettifogger0/README.md]: After I wrote the original comment above, but before I finished my review, you changed the variable name to `users_account_id`, but you forgot to update it in the sample `tfvars` file in each of the READMEs.

File: dns/README.md (hunk 2 above)
Comment (jsf9k) [reply dns/README.md]: Fixed in 45af422.

File: audit/README.md (hunk 1 above)
Comment (jsf9k) [reply audit/README.md]: Fixed in 45af422.

File: pettifogger0/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jsf9k) [reply pettifogger0/README.md]: Fixed in 45af422.

File: dns/README.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+At this point the account has been bootstrapped, and you can apply
+future changes by simply running `terraform apply
+-var-file=<workspace_name>.tfvars`.
+
+## Inputs ##
+
+| Name | Description | Type | Default | Required |
+|------|-------------|:----:|:-------:|:--------:|
+| aws_region | The AWS region where the non-global resources for this account are to be created (e.g. us-east-1). | string | `us-east-1` | no |
+| certificate_bucket_name | The name to use for the S3 bucket that will store the Certboto certificates. | string | `cisa-cool-certificates` | no |
```
[the remark this answers is not in the document]
Comment (jsf9k) [reply dns/README.md]: I decided to move the certboto Terraform resources into their own Terraform module, in keeping with the philosophy that this repo should only set up the bare minimum to get going.  I will keep this in mind when I create that other module.

What about the bucket and DynamoDB table names in Terraform?  Do you want me to change those to required inputs?

**Comment by jsf9k** [comment]
> Should we need to make it clear that the admin user accounts must already be created? Or is that too obvious?

Remedied in 04c9a30.

File: dns/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (dav3r) [reply dns/README.md]: Yes, I think any resources that must be uniquely-named should be required inputs.

[the remark this answers is not in the document]
Comment (jsf9k) [reply dns/README.md]: See 6092b7f.

File: users/README.md [hunk 5]
```diff
@@ -47,21 +41,18 @@ To do this, follow these steps:
    from the bootstrapping the terraform subdirectory:
 
    ```console
-   access_terraform_backend_role_arn = "arn:aws:iam::111111111111:role/AccessTerraformBackend"
+   this_account_id = "236526679726"
```
Comment (dav3r) [inline users/README.md]: Replace account id value with `111111111111` to match the other READMEs.

Comment (jsf9k) [reply users/README.md]: Fixed in 5386f99.

**Pull request merged**